masonry/src/widget/progress_bar.rs

```diff
@@ -0,0 +1,281 @@
+use tracing::{trace, trace_span, Span};
+use vello::Scene;
+
+use crate::kurbo::Size;
+use crate::paint_scene_helpers::{fill_lin_gradient, stroke, UnitPoint};
+use crate::text::TextLayout;
+use crate::widget::WidgetMut;
+use crate::{
+    theme, AccessCtx, AccessEvent, ArcStr, BoxConstraints, EventCtx, LayoutCtx, LifeCycle,
+    LifeCycleCtx, PaintCtx, PointerEvent, StatusChange, TextEvent, Widget, WidgetId,
+};
+
+/// A progress bar
+pub struct ProgressBar {
+    /// A value in the range `[0, 1]` inclusive, where 0 is 0% and 1 is 100% complete.
+    ///
+    /// `None` variant can be used to show a progress bar without a percentage.
+    /// It is also used if an invalid float (outside of [0, 1]) is passed.
+    progress: Option<f64>,
+    label: TextLayout<ArcStr>,
+}
+
+impl ProgressBar {
+    /// Create a new `ProgressBar`.
+    ///
+    /// `progress` is a number between 0 and 1 inclusive. If it is `NaN`, then an
+    /// indefinite progress bar will be shown.
+    /// Otherwise, the input will be clamped to [0, 1].
+    pub fn new(progress: Option<f64>) -> Self {
+        let mut out = Self::new_indefinite();
+        out.set_progress(progress);
+        out
+    }
+
+    fn new_indefinite() -> Self {
+        Self {
+            progress: None,
+            label: TextLayout::new("".into(), crate::theme::TEXT_SIZE_NORMAL as f32),
+        }
+    }
+
+    fn set_progress(&mut self, mut progress: Option<f64>) {
+        clamp_progress(&mut progress);
+        // check to see if we can avoid doing work
+        if self.progress != progress {
+            self.progress = progress;
+            self.update_text();
+        }
+    }
+
+    /// Updates the text layout with the current part-complete value
+    fn update_text(&mut self) {
+        self.label.set_text(self.value());
+    }
+
+    fn value(&self) -> ArcStr {
+        if let Some(value) = self.progress {
+            format!("{:.0}%", value * 100.).into()
+        } else {
+            "".into()
+        }
+    }
+
+    fn value_accessibility(&self) -> Box<str> {
+        if let Some(value) = self.progress {
+            format!("{:.0}%", value * 100.).into()
+        } else {
+            "progress unspecified".into()
+        }
+    }
+}
+
+// --- MARK: WIDGETMUT ---
+impl WidgetMut<'_, ProgressBar> {
+    pub fn set_progress(&mut self, progress: Option<f64>) {
+        self.widget.set_progress(progress);
+        self.ctx.request_layout();
+        self.ctx.request_accessibility_update();
+    }
+}
+
+/// Helper to ensure progress is either a number between [0, 1] inclusive, or `None`.
+///
+/// NaNs are converted to `None`.
+fn clamp_progress(progress: &mut Option<f64>) {
+    if let Some(value) = progress {
+        if value.is_nan() {
+            *progress = None;
+        } else {
+            *progress = Some(value.clamp(0., 1.));
+        }
+    }
+}
+
+// --- MARK: IMPL WIDGET ---
+impl Widget for ProgressBar {
+    // pointer events unhandled for now
```

Comment:
for now? What would you expect a pointer event on a progress bar to do?

Maybe copy the current progress percentage to the clipboard?